service/src/main/java/bio/terra/policy/service/pao/PaoService.java

```diff
@@ -102,6 +107,94 @@ public LinkSourceResult linkSourcePao(
       walker.applyChanges();
     }
 
-    return new LinkSourceResult(modifiedPao, conflicts);
+    return new PolicyUpdateResult(modifiedPao, conflicts);
+  }
+
+  /**
+   * Update the attributes of a Pao and propagate changes.
+   *
+   * @param targetPaoId the object to update
+   * @param addAttributes policy inputs to add
+   * @param removeAttributes policy inputs to remove
+   * @param updateMode how to handle applying the changes
+   */
+  public PolicyUpdateResult updatePao(
+      UUID targetPaoId,
+      PolicyInputs addAttributes,
+      PolicyInputs removeAttributes,
+      PaoUpdateMode updateMode) {
+    logger.info(
+        "UpdatePao: target {} adds {} removes {} updateMode {}",
+        targetPaoId,
+        addAttributes,
+        removeAttributes,
+        updateMode);
+
+    Pao targetPao = paoDao.getPao(targetPaoId);
+    Pao modifiedPao = targetPao.duplicateWithoutConflicts();
+    PolicyInputs newAttributes = new PolicyInputs();
+
+    // We do the removes first, so we don't remove newly added things
+    for (PolicyInput removePolicy : removeAttributes.getInputs().values()) {
+      PolicyInput existingPolicy = targetPao.getAttributes().lookupPolicy(removePolicy);
+      if (existingPolicy != null) {
+        // We have something to remove
+        PolicyInput removeResult = PolicyMutator.remove(existingPolicy, removePolicy);
+        if (removeResult != null) {
+          // There is something left of the policy to keep
+          newAttributes.addInput(removeResult);
+        }
+      }
+    }
+
+    for (PolicyInput addPolicy : addAttributes.getInputs().values()) {
+      PolicyInput existingPolicy = targetPao.getAttributes().lookupPolicy(addPolicy);
+      if (existingPolicy == null) {
+        // Nothing to combine; just take the new
+        newAttributes.addInput(addPolicy);
+      } else {
+        PolicyInput addResult = PolicyMutator.combine(existingPolicy, addPolicy);
+        if (addResult != null) {
+          // We have a combined policy to add
+          newAttributes.addInput(addResult);
+        } else {
+          throw new InvalidDirectConflictException(
+              String.format(
+                  "Update of policy %s adding %s creates a conflict",
+                  existingPolicy.getKey(), addPolicy.getKey()));
+        }
+      }
+    }
+
+    // Set the attributes to the newly computed attributes
+    // Note: we have to leave the effective attributes set to the original
+    // value so the walker detects the change.
+    modifiedPao.setAttributes(newAttributes);
+
+    // Evaluate the change, calculating new effective attribute sets and finding conflicts
+    Walker walker = new Walker(paoDao, targetPao, modifiedPao);
```

Comment:
general question, is there a race condition risk in the chuck of code before the walk?

Reply:
I don't think there is a race - it is all in a DB transaction. But I get that the Walker class has to be called in order.

Would it seem cleaner to you if I did the walk within the constructor? Then the operations on the Walker object would be `getConflicts` and `applyChanges`. That does make a kind of sense.